### PR TITLE
[Notifier] add Mailjet SMS bridge

### PR DESCRIPTION
**Pull request by jnadaud**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| License       | MIT
| Doc | https://github.com/symfony/symfony-docs/pull/15436

Adding Mailjet SMS bridge for Notifier component

### CODE REVIEW

File: src/Symfony/Component/Notifier/Bridge/Mailjet/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+Mailjet Notifier
+================
+
+Provides [Mailjet](https://mailjet.com) integration for Symfony Notifier.
+
+DSN example
+-----------
+
+```
+MAILJET_DSN=mailjet://AUTH_TOKEN@default?from=0611223344
```
Comment (94noni) [inline src/Symfony/Component/Notifier/Bridge/Mailjet/README.md]: Maybe detail that it is the sms api, not the mail api
Also what if someone add the mail api part, will it overlap?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 2]
```diff
@@ -2427,6 +2428,7 @@ private function registerNotifierConfiguration(array $config, ContainerBuilder $
             IqsmsTransportFactory::class => 'notifier.transport_factory.iqsms',
             LightSmsTransportFactory::class => 'notifier.transport_factory.lightsms',
             LinkedInTransportFactory::class => 'notifier.transport_factory.linkedin',
+            MailjetNotifierTransportFactory::class => 'notifier.transport_factory.mailjet',
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:2433]:
```suggestion
            MailjetTransportFactory::class => 'notifier.transport_factory.mailjet',
```

File: src/Symfony/Component/Notifier/Bridge/Mailjet/CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.4
+---
+
+ * Added the bridge
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Mailjet/CHANGELOG.md]:
```suggestion
 * Add the bridge
```

File: src/Symfony/Component/Notifier/Bridge/Mailjet/LICENSE [hunk 4]
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2019-2021 Fabien Potencier
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Mailjet/LICENSE]:
```suggestion
Copyright (c) 2021 Fabien Potencier
```

File: src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+use Symfony\Component\Notifier\Exception\UnsupportedMessageTypeException;
+use Symfony\Component\Notifier\Message\MessageInterface;
+use Symfony\Component\Notifier\Message\SentMessage;
+use Symfony\Component\Notifier\Message\SmsMessage;
+use Symfony\Component\Notifier\Transport\AbstractTransport;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php]: Is this true?

File: src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        if (!$message instanceof SmsMessage) {
+            throw new UnsupportedMessageTypeException(__CLASS__, SmsMessage::class, $message);
+        }
+
+        $endpoint = sprintf('https://%s/v4/sms-send', $this->getEndpoint());
+
+        $response = $this->client->request('POST', $endpoint, [
+            'headers' => [
+                'Authorization' => 'Bearer '.$this->authToken,
+            ],
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php]:
```suggestion
            'auth_bearer' => $this->authToken,
```

File: src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransportFactory.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace Symfony\Component\Notifier\Bridge\Mailjet;
+
+use Symfony\Component\Notifier\Exception\UnsupportedSchemeException;
+use Symfony\Component\Notifier\Transport\AbstractTransportFactory;
+use Symfony\Component\Notifier\Transport\Dsn;
+use Symfony\Component\Notifier\Transport\TransportInterface;
+
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransportFactory.php]: Is this true?

File: src/Symfony/Component/Notifier/Bridge/Mailjet/README.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+```
+
+where:
+ - `AUTH_TOKEN` is your Mailjet SMS auth token
+ - `FROM` is the sender
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Mailjet/README.md]: Is it a string? Is it a phone number?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 2 above)
Comment (jnadaud) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:2431]: There is already an MailjetTransportFactory used by Symfony Mailer component

Comment (jnadaud) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:2433]: There is already an MailjetTransportFactory used by Symfony Mailer component

File: src/Symfony/Component/Notifier/Bridge/Mailjet/README.md (hunk 1 above)
Comment (jnadaud) [reply src/Symfony/Component/Notifier/Bridge/Mailjet/README.md]: Mail api part is already used in Mailjet Symfony Mailer Bridge, and it used MAILER_DSN as suggested env variable

File: src/Symfony/Component/Notifier/Bridge/Mailjet/LICENSE (hunk 4 above)
Comment (jnadaud) [reply src/Symfony/Component/Notifier/Bridge/Mailjet/LICENSE]: Done

File: src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php (hunk 5 above)
Comment (jnadaud) [reply src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php]: Done

File: src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php (hunk 6 above)
Comment (jnadaud) [reply src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php]: Done

File: src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransportFactory.php (hunk 7 above)
Comment (jnadaud) [reply src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransportFactory.php]: Done

File: src/Symfony/Component/Notifier/Bridge/Mailjet/README.md (hunk 8 above)
Comment (jnadaud) [reply src/Symfony/Component/Notifier/Bridge/Mailjet/README.md]: Done

**Comment by OskarStark** [comment]
It looks, like `from` is required and `user`/`password` is not used, I would use `from` as `user` and `authtoken` as `pass` for the DSN, like explained here: https://github.com/symfony/symfony/pull/41522#discussion_r644752617

**Comment by jnadaud** [comment]
@OskarStark Done!!!

**Review by OskarStark — APPROVED** [review]

File: src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        $this->authToken = $authToken;
+        $this->from = $from;
+
+        parent::__construct($client, $dispatcher);
+    }
+
+    public function __toString(): string
+    {
+        return sprintf('mailjet://%s:%s@%s', $this->from, $this->authToken, $this->getEndpoint());
```
Comment (OskarStark) [inline src/Symfony/Component/Notifier/Bridge/Mailjet/MailjetTransport.php]: The auth-token should not be exposed, please remove it and fix the tests, thanks

**Comment by fabpot** [comment]
@jnadaud Can you make the last change asked by Oskar? Thank you.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jnadaud.

**Comment by fabpot** [comment]
@jnadaud Can you create a PR on symfony/recipes for this new notifier?

**Comment by nicolas-grekas** [comment]
Tests fail also after this merge. Could you please have a look @jnadaud?

**Comment by OskarStark** [comment]
I am on it 👍 

https://github.com/symfony/symfony/pull/41985

**Pull request merged**